### PR TITLE
add log out button

### PR DESCRIPTION
**Pull request by AntonioLok**
# Description

This PR implements the follow:
1 - When user is authenticated, the log out button is displayed. If a token has expired, the token is removed from the local storage.

# Test

#### Backend
Set-up backend server and start the server
0 - Checkout master branch `git checkout master`.
1 - Install latest dependencies `npm i`.
2 - Set the required environment variable `JWT_SECRET` by creating a `.env` file in the main directory and set your secret in that file ie. `JWT_SECRET=HELLO`.
3 - run the app using `npm run start`. If you want to skip step 2, you can set the env with `JWT_SECRET=HELLO npm run start`.

#### Front-end
Set-up frontend webapp and start the app:
0 - Checkout this feature branch `git checkout add-log-out-button`.
1 - Install latest dependencies `npm i`.
2 - run the app using `npm run start`.

### 2- After the server and web app have been started, navigate to the login page by clicking on the login button. Login with your credentials that you used last time or use the following one:
```
email: test@gmail.com
password: test@gmail.com
```

Upon logging in, the `logout` button should now appear.

### CODE REVIEW

**Comment by AntonioLok** [comment]
I know you have been very busy but just in case you forgot about this PR... :). @howardhenry

File: src/components/app.js [hunk 1]
```diff
@@ -17,20 +21,56 @@ const {
...
-      </Switch>
-    </div>
-  </Router>
-);
-
-export default App;
+class App extends Component {
+  shouldComponentUpdate(nextProps) {
+    const { loginState } = this.props;
+    // deep comparisons
```
Comment (howardhenry) [inline src/components/app.js]: You can remove this comment as the function name should be descriptive enough.

File: src/components/app.js [hunk 2]
```diff
@@ -17,20 +21,56 @@ const {
...
+
+    let isAuthenticated = false;
+    if (token) {
+      const decoded = jwtDecode(token);
+      if (Date.now() < decoded.exp * 1000) {
+        isAuthenticated = true;
+      } else {
+        localStorage.removeItem('OS_AUTH_TOKEN');
+      }
+    }
```
Comment (howardhenry) [inline src/components/app.js]: The logic to check if the use is authenticated doesn't belong inside a component. This should be somewhere like an authentication helper to ensure separation of concerns and reusability.

File: src/components/common/nav-menu/index.js [hunk 3]
```diff
@@ -1,13 +1,32 @@
...
 import { Link } from 'react-router-dom';
+import PropTypes from 'prop-types';
 import { ROUTES } from '../../../constants';
 
 class NavMenu extends PureComponent {
+  // eslint-disable-next-line class-methods-use-this
+  logOut() {
+    localStorage.removeItem('OS_AUTH_TOKEN');
+  }
+
```
Comment (howardhenry) [inline src/components/common/nav-menu/index.js]: Like the authentication check, the log out function should not be defined in a component. Move this to an authentication helper/util as you will likely reuse the function again in a different component or module.

**Comment by AntonioLok** [comment]
> So sorry I completely missed this one!
> 
> It looks great! I've just added a few comments to keep up best practices.

No worries. Thanks for the feedback. I have updated the PR as per comments. Thanks for your time!

**Pull request merged**